## Assignment 2 Numpy and pandas

This assignment will contain 3 questions with details as below. The due date is October 5 (Friday), 2018 23:59PM. Each late day will result in 20% loss of total points.

### Question 1 (30 points) Numpy is fast!

Suppose we need to compute the cumulative sum of $\sum_{i=0}^n \alpha^i$ for given $\alpha$ and $n$. 

For example, when $\alpha=0.5$ and $n=10$, the cumulative sum of $\sum_{i=0}^{10} 0.5^i$ returns `[1.0, 1.5, 1.75, 1.875, 1.9375, 1.96875, 1.984375, 1.9921875, 1.99609375, 1.998046875, 1.9990234375]`


As a courtesy, I implement the following function `cum_sum` that can be used to generate a list of cumulative sum when iterating through a range generator.

In [1]:
def cum_sum(alpha, n):
    current = 1.0
    sum = current
    for i in range(n):
        current = current * alpha
        sum = sum + current
    return sum 

cumsum = []
for i in range(11):
    cumsum.append(cum_sum(0.5, i))
    
print(cumsum)

[1.0, 1.5, 1.75, 1.875, 1.9375, 1.96875, 1.984375, 1.9921875, 1.99609375, 1.998046875, 1.9990234375]


We can calculate how much time does it spend to run this code using `time` module as below:

In [2]:
import time

begin = time.time()
n_samples = 100000

cumsum = []
for i in range(n_samples):
    cumsum.append(cum_sum(0.5, i))
print(cumsum)
    
end = time.time()

time0 = end-begin
print("Time took to run: {} seconds.".format(time0))

[1.0, 1.5, 1.75, 1.875, 1.9375, 1.96875, 1.984375, 1.9921875, 1.99609375, 1.998046875, 1.9990234375, 1.99951171875, 1.999755859375, 1.9998779296875, 1.99993896484375, 1.999969482421875, 1.9999847412109375, 1.9999923706054688, 1.9999961853027344, 1.9999980926513672, 1.9999990463256836, 1.9999995231628418, 1.999999761581421, 1.9999998807907104, 1.9999999403953552, 1.9999999701976776, 1.9999999850988388, 1.9999999925494194, 1.9999999962747097, 1.9999999981373549, 1.9999999990686774, 1.9999999995343387, 1.9999999997671694, 1.9999999998835847, 1.9999999999417923, 1.9999999999708962, 1.999999999985448, 1.999999999992724, 1.999999999996362, 1.999999999998181, 1.9999999999990905, 1.9999999999995453, 1.9999999999997726, 1.9999999999998863, 1.9999999999999432, 1.9999999999999716, 1.9999999999999858, 1.999999999999993, 1.9999999999999964, 1.9999999999999982, 1.9999999999999991, 1.9999999999999996, 1.9999999999999998, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 

It takes about 355 seconds on my machine to run the code with 10,000 samples. Note that this time may vary depending on the memory and CPU of your machine. 

**Question 1.1** (15 points) Now implement a list comprehension for the same purpose and estimate how much time does it take to generate a list of cumulative sum for 10,000 samples

Hint: you can use the method accumulate in the module itertools. Check the documentation of itertools at [here](https://docs.python.org/3/library/itertools.html#itertools.accumulate)

In [3]:
# Question 1

import itertools

begin = time.time()
n_samples = 100000

# write your code here
cumsum = list(itertools.accumulate([0.5**i for i in range(n_samples)]))

print(cumsum)
end = time.time()

time1 = end-begin

print("Time took to run: {} seconds.".format(time1))

[1.0, 1.5, 1.75, 1.875, 1.9375, 1.96875, 1.984375, 1.9921875, 1.99609375, 1.998046875, 1.9990234375, 1.99951171875, 1.999755859375, 1.9998779296875, 1.99993896484375, 1.999969482421875, 1.9999847412109375, 1.9999923706054688, 1.9999961853027344, 1.9999980926513672, 1.9999990463256836, 1.9999995231628418, 1.999999761581421, 1.9999998807907104, 1.9999999403953552, 1.9999999701976776, 1.9999999850988388, 1.9999999925494194, 1.9999999962747097, 1.9999999981373549, 1.9999999990686774, 1.9999999995343387, 1.9999999997671694, 1.9999999998835847, 1.9999999999417923, 1.9999999999708962, 1.999999999985448, 1.999999999992724, 1.999999999996362, 1.999999999998181, 1.9999999999990905, 1.9999999999995453, 1.9999999999997726, 1.9999999999998863, 1.9999999999999432, 1.9999999999999716, 1.9999999999999858, 1.999999999999993, 1.9999999999999964, 1.9999999999999982, 1.9999999999999991, 1.9999999999999996, 1.9999999999999998, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 

In [4]:
time0/time1

9890.706274891818

**Question 1.2** (15 points) Now implement using numpy for the same purpose and estimate how much time does it take to generate a list of cumulative sum for 10,000 samples (in order to receive full score, your program must be at least 1500 times faster than the for loop)

You may receive 5 bonus points if your program is at least 5000 times faster than the for loop!

In [5]:
import numpy as np

In [6]:
begin = time.time()
n_samples = 100000
alpha = 0.5
# write your code here
np.cumsum([alpha**np.arange(n_samples)])

print(cumsum)
end = time.time()

time2 = end-begin

print("Time took to run: {} seconds.".format(time2))

[1.0, 1.5, 1.75, 1.875, 1.9375, 1.96875, 1.984375, 1.9921875, 1.99609375, 1.998046875, 1.9990234375, 1.99951171875, 1.999755859375, 1.9998779296875, 1.99993896484375, 1.999969482421875, 1.9999847412109375, 1.9999923706054688, 1.9999961853027344, 1.9999980926513672, 1.9999990463256836, 1.9999995231628418, 1.999999761581421, 1.9999998807907104, 1.9999999403953552, 1.9999999701976776, 1.9999999850988388, 1.9999999925494194, 1.9999999962747097, 1.9999999981373549, 1.9999999990686774, 1.9999999995343387, 1.9999999997671694, 1.9999999998835847, 1.9999999999417923, 1.9999999999708962, 1.999999999985448, 1.999999999992724, 1.999999999996362, 1.999999999998181, 1.9999999999990905, 1.9999999999995453, 1.9999999999997726, 1.9999999999998863, 1.9999999999999432, 1.9999999999999716, 1.9999999999999858, 1.999999999999993, 1.9999999999999964, 1.9999999999999982, 1.9999999999999991, 1.9999999999999996, 1.9999999999999998, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 

In [7]:
time0/time2

18781.140669435008

We know that $\sum_{i=0}^{n} \alpha^i = \frac{1-\alpha^{(n+1)}}{1-\alpha} $, so we perform element-wise operations as:

In [10]:
begin = time.time()
n_samples = 100000
alpha = 0.5
# write your code here
cumsum = 1+(1.0-0.5**(np.arange(n_samples)+1)/0.5)

print(cumsum)
end = time.time()

time3 = end-begin

print("Time took to run: {} seconds.".format(time3))

[1.   1.5  1.75 ... 2.   2.   2.  ]
Time took to run: 0.0021517276763916016 seconds.


In [11]:
time0/time3

165378.08853185596

### Question 2 (30 points) Monte Carlo 

Monte Carlo is a city in Monacco where the famous Monte Carlo casino is located.

![casino](http://www.casinomontecarlo.com/wp-content/uploads/2017/06/casino-de-monte-carlo-1100x358.jpg)

In light of this, Monte Carlo methods (or Monte Carlo experiments) are a broad class of computational algorithms that rely on repeated random sampling to obtain numerical results. Their essential idea is using randomness to solve problems that might be deterministic in principle. They are often used in physical and mathematical problems and are most useful when it is difficult or impossible to use other approaches. Monte Carlo methods are mainly used in three problem classes: optimization, numerical integration, and generating draws from a probability distribution.

**Estimate the Pi**

In order to estimate the $\pi$, the idea is to simulate random (x, y) points in a 2-D plane with domain as a square of side 1 unit. Imagine a circle inside the same domain with same diameter and inscribed into the square. We can generate a large number of uniformly distributed random points and plot them on the graph. These points can be in any position within the square i.e. between (0,0) and (1,1). We keep track of the total number of points, and the number of points that are inside the circle. If we divide the number of points within the circle, $N_{inner}$ by the total number of points, $N_{total}$, we should get a value that is an approximation of the ratio of the areas we calculated above, $\pi/4$.

Write a function `approximate_pi` with argument `number_simulations` to approximate the Pi value using Monte Carlo simulations. You may consider to use `numpy.random` to make random draws. 

Give a rough estimate about how many random draws you may need to achieve accuray of 99.999% (by comparing with numpy.pi).

In [17]:
# Question 2


def approximate_pi(number_simulations):
    number_inner = 0
    number_total = number_simulations
    for i in range(number_simulations):
        x, y = np.random.uniform(0, 1, size=2)
        r = (x**2 + y**2)**0.5
        if r < 1:
            number_inner += 1
            
    pi = number_inner/number_total*4        
    return pi


In [23]:
number_simulations = 1000000

estimated_pi  = approximate_pi(number_simulations)
error = np.abs((estimated_pi - np.pi) / np.pi)
accuracy = 1 - error

print('The accuracy with {} simulations is: {}'.format(number_simulations, accuracy))

The accuracy with 1000000 simulations is: 0.9999543714204823


### Question 3 (40 points) California housing

We will explore the famous Califronia housing dataset. The original database is available from StatLib http://lib.stat.cmu.edu/datasets/


The data contains 20,640 observations on 9 variables.
This dataset contains the average house value as target variable
and the following input variables (features): records
the following for each tract in California: Median house price, median
house age, average number of rooms per house, average number of bedrooms,
average number of occupants, total number of houses, median income
(in thousands of dollars), latitude and longitude.

You can download the data from the Internet as:

```python
from sklearn.datasets.california_housing import fetch_california_housing

cal_housing = fetch_california_housing()

```

The dataset has the following format:
```
    dataset : dict-like object with the following attributes:
    dataset.data : ndarray, shape [20640, 8]
        Each row corresponding to the 8 feature values in order.
    dataset.target : numpy array of shape (20640,)
        Each value corresponds to the average house value in units of 100,000.
    dataset.feature_names : array of length 8
        Array of ordered feature names used in the dataset.
    dataset.DESCR : string
        Description of the California housing dataset.
```

** Question 3.1 ** (5 points) Create a dataframe of `california_housing` from the fetched dataset using the data, target and feature_names, shows how many observations/features

In [24]:
import pandas as pd

In [25]:
housing = pd.read_csv('housing.csv')

In [26]:
housing.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

In [27]:
housing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
longitude             20640 non-null float64
latitude              20640 non-null float64
housing_median_age    20640 non-null float64
total_rooms           20640 non-null float64
total_bedrooms        20433 non-null float64
population            20640 non-null float64
households            20640 non-null float64
median_income         20640 non-null float64
median_house_value    20640 non-null float64
ocean_proximity       20640 non-null object
dtypes: float64(9), object(1)
memory usage: 1.6+ MB


In [28]:
print('There are in total {} observations and {} features'.format(housing.shape[0], housing.shape[1]))

There are in total 20640 observations and 10 features


** Question 3.2** (5 points) Show the descripive statistics of features

Using `pandas.describe`() can easily show the descriptive statistics of features with numerical values

In [29]:
housing.describe()

longitude      latitude  housing_median_age   total_rooms  \
count  20640.000000  20640.000000        20640.000000  20640.000000   
mean    -119.569704     35.631861           28.639486   2635.763081   
std        2.003532      2.135952           12.585558   2181.615252   
min     -124.350000     32.540000            1.000000      2.000000   
25%     -121.800000     33.930000           18.000000   1447.750000   
50%     -118.490000     34.260000           29.000000   2127.000000   
75%     -118.010000     37.710000           37.000000   3148.000000   
max     -114.310000     41.950000           52.000000  39320.000000   

       total_bedrooms    population    households  median_income  \
count    20433.000000  20640.000000  20640.000000   20640.000000   
mean       537.870553   1425.476744    499.539680       3.870671   
std        421.385070   1132.462122    382.329753       1.899822   
min          1.000000      3.000000      1.000000       0.499900   
25%        296.000000    787.000000    280.000000       2.563400   
50%        435.000000   1166.000000    409.000000       3.534800   
75%        647.000000   1725.000000    605.000000       4.743250   
max       6445.000000  35682.000000   6082.000000      15.000100   

       median_house_value  
count        20640.000000  
mean        206855.816909  
std         115395.615874  
min          14999.000000  
25%         119600.000000  
50%         179700.000000  
75%         264725.000000  
max         500001.000000

** Question 3.3** (15 points) Compare the areas with houses having more than 25 years old with areas with houses having less than 10 years old, what do you find about their total rooms, total bedrooms, median household incomes?

First, we can filter the rows and columns to obtain the required subset of data:

In [31]:
old_housing_area = housing.loc[housing['housing_median_age'] > 25, ['total_rooms', 'total_bedrooms', 'median_income']]
new_housing_area = housing.loc[housing['housing_median_age'] < 10, ['total_rooms', 'total_bedrooms', 'median_income']]

Then we can calculate the difference in means between two DataFrames:

In [34]:
new_housing_area.mean() - old_housing_area.mean()

total_rooms       2907.662981
total_bedrooms     475.768563
median_income        0.947114
dtype: float64

We find that on average, areas with houses having less than 10 years old (new housing areas) generally have 2908 more total_rooms, and 476 more bedrooms, compared with areas with houses having more than 25 years old (old housing areas). Meanwhile, median income of new housing areas is $9471 more than that of old housing areas.

In [35]:
new_housing_area.median() - old_housing_area.median()

total_rooms       1638.0000
total_bedrooms     304.0000
median_income        0.9167
dtype: float64

You can also calculate difference in median as above. How do you compare this with difference in means?

** Question 3.4 ** (15 points) What is the difference between average median house price for households with top 20% of median houshold income and bottom 20% of median household income?

Let's first derive the top/bottom 20% of median housing income using `pandas.quantile()`

In [44]:
top_income = housing['median_income'].quantile(0.8)
bottom_income = housing['median_income'].quantile(0.2)

Then filter out areas in these two categories:

In [46]:
top_income_area = housing.loc[housing['median_income'] >= top_income]
bottom_income_area = housing.loc[housing['median_income'] <= bottom_income]

In [54]:
mean_difference = top_income_area['median_house_value'].mean() - bottom_income_area['median_house_value'].mean()
print('The difference between average median house price for households with top 20% of'
      'median houshold income and bottom 20% of median household income is ${}'.format(mean_difference))

The difference between average median house price for households with top 20% ofmedian houshold income and bottom 20% of median household income is $216588.25762606284
